nft/ethereum_payment_obligation.go

```diff
@@ -80,45 +88,155 @@ func newEthereumPaymentObligation(
 	}
 }
 
-func (s *ethereumPaymentObligation) prepareMintRequest(ctx context.Context, tokenID TokenID, documentID []byte, depositAddress string, proofFields []string) (MintRequest, error) {
-	model, err := s.docSrv.GetCurrentVersion(ctx, documentID)
+func (s *ethereumPaymentObligation) prepareMintRequest(ctx context.Context, tokenID TokenID, cid identity.CentID, req MintNFTRequest) (mreq MintRequest, err error) {
+	model, err := s.docSrv.GetCurrentVersion(ctx, req.DocumentID)
 	if err != nil {
-		return MintRequest{}, err
+		return mreq, err
 	}
 
 	corDoc, err := model.PackCoreDocument()
 	if err != nil {
-		return MintRequest{}, err
+		return mreq, err
 	}
 
-	proofs, err := s.docSrv.CreateProofs(ctx, documentID, proofFields)
+	docProofs, err := s.docSrv.CreateProofs(ctx, req.DocumentID, req.ProofFields)
 	if err != nil {
-		return MintRequest{}, err
+		return mreq, err
+	}
+
+	nftProofs, err := generateNFTProofs(corDoc, req, tokenID, cid)
+	if err != nil {
+		return mreq, err
+	}
+
+	var pfs []*proofspb.Proof
+	pfs = append(pfs, docProofs.FieldProofs...)
+	for _, p := range nftProofs {
+		p := p
+		pfs = append(pfs, &p)
 	}
 
-	toAddress := common.HexToAddress(depositAddress)
+	toAddress := common.HexToAddress(req.DepositAddress)
 
 	anchorID, err := anchors.ToAnchorID(corDoc.CurrentVersion)
 	if err != nil {
-		return MintRequest{}, err
+		return mreq, err
 	}
 
 	rootHash, err := anchors.ToDocumentRoot(corDoc.DocumentRoot)
 	if err != nil {
-		return MintRequest{}, err
+		return mreq, err
 	}
 
-	requestData, err := NewMintRequest(tokenID, toAddress, anchorID, proofs.FieldProofs, rootHash)
+	requestData, err := NewMintRequest(tokenID, toAddress, anchorID, pfs, rootHash)
 	if err != nil {
-		return MintRequest{}, err
+		return mreq, err
 	}
 
 	return requestData, nil
 
 }
 
+// generateNFTProofs generates required NFT uniqueness proofs
+func generateNFTProofs(cd *coredocumentpb.CoreDocument, req MintNFTRequest, tokenID TokenID, cid identity.CentID) ([]proofspb.Proof, error) {
+	cdTree, err := coredocument.GetCoreDocTree(cd)
+	if err != nil {
+		return nil, err
+	}
+
+	var pfs []proofspb.Proof
+	registry := common.HexToAddress(req.RegistryAddress)
+	if req.SubmitTokenProof {
+		pf, err := createTokenProof(cd, cdTree, registry)
+		if err != nil {
+			return nil, err
+		}
+
+		pfs = append(pfs, pf)
+	}
+
+	if req.GrantNFTReadAccess && req.SubmitNFTReadAccessProof {
+		pf, err := createNFTReadAccessProof(cd, cdTree, registry, tokenID[:])
+		if err != nil {
+			return nil, err
+		}
+
+		pfs = append(pfs, pf...)
+	}
+
+	if req.SubmitRoleProof != "" {
+		pf, err := createRoleProof(cd, cdTree, req.SubmitRoleProof, cid)
+		if err != nil {
+			return nil, err
+		}
+
+		pfs = append(pfs, pf)
+	}
+
```

Comment:
I think a proof is missing here for the field `NextVersion`. We should probably also add an option to the MintRequest to only add it if required. I think @rdinicut needs it to verify the document is the newest version before minting the NFT.

Reply:
It should be part of the default proofs fields that get sent by the client in the request. My understanding is that it was the choice we made instead of adding specific proofs by default in the Mint method

Reply:
Ok, you are right we can just supply it as a field.